File: library/ilib/pointers/ffi/pointers_ffi.go
// AC ilib: pointers — Go FFI
// Pointer storage and operations

package pointers

import (
	"sync"
)

// ============================================================================
// POINTER REGISTRY
// ============================================================================

type PtrEntry struct {
	addr    interface{} // actual reference/value
	typeID  int64       // type hint (0=generic)
	valid   bool        // is valid?
	data    []byte      // allocated buffer (if applicable)
}

var (
	ptrRegistry = make(map[int64]*PtrEntry)
	nextPtrID   int64 = 1000
	registryMu  sync.RWMutex
	NULL_PTR    int64 = -1
)

// ============================================================================
// POINTER CREATION
// ============================================================================

// PtNew - Create pointer to a value
func PtNew(value interface{}, valueSize int64, typeID int64) int64 {
	if value == nil || valueSize <= 0 {
		return NULL_PTR
	}

	registryMu.Lock()
	defer registryMu.Unlock()

	ptrID := nextPtrID
	nextPtrID++
	ptrRegistry[ptrID] = &PtrEntry{
		addr:   value,
		typeID: typeID,
		valid:  true,
	}
	return ptrID
}

// PtAlloc - Allocate storage of given size
func PtAlloc(valueSize int64, typeID int64) int64 {
	if valueSize <= 0 {
		return NULL_PTR
	}

	registryMu.Lock()
	defer registryMu.Unlock()

	ptrID := nextPtrID
	nextPtrID++
	ptrRegistry[ptrID] = &PtrEntry{
		addr:   make([]byte, valueSize),
		typeID: typeID,
		valid:  true,
		data:   make([]byte, valueSize),
	}
	return ptrID
}

// ============================================================================
// POINTER DEREFERENCING
// ============================================================================

// PtGet - Dereference pointer - get value
func PtGet(ptrID int64) interface{} {
	registryMu.RLock()
	defer registryMu.RUnlock()

	entry, ok := ptrRegistry[ptrID]
	if !ok || !entry.valid {
		return nil
	}
	return entry.addr
}

// PtSet - Set value at pointer location
func PtSet(ptrID int64, value interface{}, offset int64) int {
	registryMu.Lock()
	defer registryMu.Unlock()

	entry, ok := ptrRegistry[ptrID]
	if !ok || !entry.valid {
		return -1
	}

	if data, isBytes := entry.addr.([]byte); isBytes {
		if num, isInt := value.(int); isInt && offset >= 0 && offset < int64(len(data)) {
			data[offset] = byte(num & 0xFF)
		} else if bytes, isByte := value.([]byte); isByte {
			copy(data[offset:], bytes)
		}
		return 0
	}

	entry.addr = value
	return 0
}

// ============================================================================
// POINTER ARITHMETIC
// ============================================================================

// PtAdd - Add offset to pointer
func PtAdd(ptrID int64, offset int64) int64 {
	registryMu.Lock()
	defer registryMu.Unlock()

	entry, ok := ptrRegistry[ptrID]
	if !ok || !entry.valid {
		return NULL_PTR
	}

	newID := nextPtrID
	nextPtrID++
	ptrRegistry[newID] = &PtrEntry{
		addr:   entry.addr,
		typeID: entry.typeID,
		valid:  true,
	}
	return newID
}

// PtSub - Subtract two pointers (get distance)
func PtSub(ptrID1, ptrID2 int64) int64 {
	registryMu.RLock()
	defer registryMu.RUnlock()

	entry1, ok1 := ptrRegistry[ptrID1]
	entry2, ok2 := ptrRegistry[ptrID2]

	if !ok1 || !ok2 || !entry1.valid || !entry2.valid {
		return 0
	}

	return 0 // Can't compute real offset between different objects
}

// ============================================================================
// POINTER COMPARISON
// ============================================================================

// PtEq - Check if two pointers are equal
func PtEq(ptrID1, ptrID2 int64) int {
	registryMu.RLock()
	defer registryMu.RUnlock()

	entry1, ok1 := ptrRegistry[ptrID1]
	entry2, ok2 := ptrRegistry[ptrID2]

	if !ok1 || !ok2 || !entry1.valid || !entry2.valid {
		return 0
	}

	if entry1.addr == entry2.addr {
		return 1
	}
	return 0
}

// PtNull - Get NULL pointer value
func PtNull() int64 {
	return NULL_PTR
}

// PtIsNull - Check if pointer is NULL
func PtIsNull(ptrID int64) int {
	if ptrID == NULL_PTR {
		return 1
	}
	return 0
}

// ============================================================================
// POINTER LIFECYCLE
// ============================================================================

// PtFree - Free/invalidate pointer
func PtFree(ptrID int64) int {
	registryMu.Lock()
	defer registryMu.Unlock()

	entry, ok := ptrRegistry[ptrID]
	if ok {
		entry.valid = false
		delete(ptrRegistry, ptrID)
		return 0
	}
	return -1
}

// PtTypeof - Get type ID of pointer
func PtTypeof(ptrID int64) int64 {
	registryMu.RLock()
	defer registryMu.RUnlock()

	entry, ok := ptrRegistry[ptrID]
	if !ok {
		return -1
	}
	if !entry.valid {
		return -1
	}
	return entry.typeID
}

// PtValid - Check if pointer is valid
func PtValid(ptrID int64) int {
	registryMu.RLock()
	defer registryMu.RUnlock()

	entry, ok := ptrRegistry[ptrID]
	if !ok {
		return 0
	}
	if entry.valid {
		return 1
	}
	return 0
}

// ============================================================================
// UTILITY
// ============================================================================

// PtSizeof - Get size of allocated memory
func PtSizeof(ptrID int64) int64 {
	registryMu.RLock()
	defer registryMu.RUnlock()

	entry, ok := ptrRegistry[ptrID]
	if !ok {
		return 0
	}

	if data, isBytes := entry.addr.([]byte); isBytes {
		return int64(len(data))
	}
	return 0
}

// PtClearAll - Clear entire registry (for cleanup)
func PtClearAll() {
	registryMu.Lock()
	defer registryMu.Unlock()

	ptrRegistry = make(map[int64]*PtrEntry)
	nextPtrID = 1000
}

File: library/ilib/pointers/ffi/pointers_ffi.js
// AC ilib: pointers — JavaScript FFI
// Pointer storage and operations

// ============================================================================
// POINTER REGISTRY (global storage)
// ============================================================================

class PtrEntry {
    constructor(addr, typeId = 0) {
        this.addr = addr;      // actual reference/value
        this.typeId = typeId;  // type hint (0=generic)
        this.valid = true;
    }
}

const ptrRegistry = new Map();  // Map<int, PtrEntry>
let nextPtrId = 1000;            // start IDs at 1000
const NULL_PTR = -1;

// ============================================================================
// POINTER CREATION
// ============================================================================

/**
 * pt_new - Create pointer to a value
 * @param {*} value - value to store
 * @param {number} valueSize - size in bytes (for type hinting)
 * @param {number} typeId - optional type identifier (0=generic)
 * @returns {number} pointer ID, or NULL_PTR on error
 */
function ptNew(value, valueSize, typeId = 0) {
    if (value === null || value === undefined || valueSize <= 0) {
        return NULL_PTR;
    }

    try {
        const ptrId = nextPtrId++;
        ptrRegistry.set(ptrId, new PtrEntry(value, typeId));
        return ptrId;
    } catch (e) {
        return NULL_PTR;
    }
}

/**
 * pt_alloc - Allocate storage of given size
 * @param {number} valueSize - size in bytes
 * @param {number} typeId - optional type identifier
 * @returns {number} pointer ID, or NULL_PTR
 */
function ptAlloc(valueSize, typeId = 0) {
    if (valueSize <= 0) return NULL_PTR;

    try {
        const ptrId = nextPtrId++;
        const buffer = Buffer.alloc(valueSize);
        ptrRegistry.set(ptrId, new PtrEntry(buffer, typeId));
        return ptrId;
    } catch (e) {
        return NULL_PTR;
    }
}

// ============================================================================
// POINTER DEREFERENCING
// ============================================================================

/**
 * pt_get - Dereference pointer - get value
 * @param {number} ptrId - pointer ID
 * @returns {*} stored value or null
 */
function ptGet(ptrId) {
    if (!ptrRegistry.has(ptrId)) return null;
    const entry = ptrRegistry.get(ptrId);
    if (!entry.valid) return null;
    return entry.addr;
}

/**
 * pt_set - Set value at pointer location
 * @param {number} ptrId - pointer ID
 * @param {*} value - new value
 * @param {number} offset - offset in bytes
 * @returns {number} 0 on success, -1 on error
 */
function ptSet(ptrId, value, offset = 0) {
    if (!ptrRegistry.has(ptrId)) return -1;
    const entry = ptrRegistry.get(ptrId);
    if (!entry.valid) return -1;

    try {
        if (Buffer.isBuffer(entry.addr)) {
            if (typeof value === 'number') {
                entry.addr[offset] = value & 0xFF;
            } else if (Buffer.isBuffer(value)) {
                value.copy(entry.addr, offset);
            } else if (typeof value === 'string') {
                entry.addr.write(value, offset);
            }
            return 0;
        } else {
            ptrRegistry.get(ptrId).addr = value;
            return 0;
        }
    } catch (e) {
        return -1;
    }
}

// ============================================================================
// POINTER ARITHMETIC
// ============================================================================

/**
 * pt_add - Add offset to pointer
 * @param {number} ptrId - pointer ID
 * @param {number} offset - bytes to add
 * @returns {number} new pointer ID, or NULL_PTR
 */
function ptAdd(ptrId, offset) {
    if (!ptrRegistry.has(ptrId)) return NULL_PTR;
    const entry = ptrRegistry.get(ptrId);
    if (!entry.valid) return NULL_PTR;

    try {
        const newId = nextPtrId++;
        ptrRegistry.set(newId, new PtrEntry(entry.addr, entry.typeId));
        return newId;
    } catch (e) {
        return NULL_PTR;
    }
}

/**
 * pt_sub - Subtract two pointers (get distance)
 * @param {number} ptrId1 - first pointer ID
 * @param {number} ptrId2 - second pointer ID
 * @returns {number} difference in bytes
 */
function ptSub(ptrId1, ptrId2) {
    if (!ptrRegistry.has(ptrId1) || !ptrRegistry.has(ptrId2)) return 0;

    const entry1 = ptrRegistry.get(ptrId1);
    const entry2 = ptrRegistry.get(ptrId2);

    if (!entry1.valid || !entry2.valid) return 0;

    if (entry1.addr === entry2.addr) return 0;
    return 0;  // Can't compute real offset between different objects
}

// ============================================================================
// POINTER COMPARISON
// ============================================================================

/**
 * pt_eq - Check if two pointers are equal
 * @returns {number} 1 if equal, 0 otherwise
 */
function ptEq(ptrId1, ptrId2) {
    if (!ptrRegistry.has(ptrId1) || !ptrRegistry.has(ptrId2)) return 0;

    const entry1 = ptrRegistry.get(ptrId1);
    const entry2 = ptrRegistry.get(ptrId2);

    if (!entry1.valid || !entry2.valid) return 0;

    return entry1.addr === entry2.addr ? 1 : 0;
}

/**
 * pt_null - Get NULL pointer value
 */
function ptNull() {
    return NULL_PTR;
}

/**
 * pt_is_null - Check if pointer is NULL
 * @returns {number} 1 if NULL, 0 otherwise
 */
function ptIsNull(ptrId) {
    return ptrId === NULL_PTR ? 1 : 0;
}

// ============================================================================
// POINTER LIFECYCLE
// ============================================================================

/**
 * pt_free - Free/invalidate pointer
 * @param {number} ptrId - pointer ID to free
 * @returns {number} 0 on success
 */
function ptFree(ptrId) {
    if (ptrRegistry.has(ptrId)) {
        ptrRegistry.get(ptrId).valid = false;
        ptrRegistry.delete(ptrId);
        return 0;
    }
    return -1;
}

/**
 * pt_typeof - Get type ID of pointer
 * @returns {number} type_id, or -1 if invalid
 */
function ptTypeof(ptrId) {
    if (!ptrRegistry.has(ptrId)) return -1;
    const entry = ptrRegistry.get(ptrId);
    return entry.valid ? entry.typeId : -1;
}

/**
 * pt_valid - Check if pointer is valid
 * @returns {number} 1 if valid, 0 otherwise
 */
function ptValid(ptrId) {
    if (!ptrRegistry.has(ptrId)) return 0;
    return ptrRegistry.get(ptrId).valid ? 1 : 0;
}

// ============================================================================
// UTILITY
// ============================================================================

/**
 * pt_sizeof - Get size of allocated memory
 * @returns {number} size in bytes, or 0 if unknown
 */
function ptSizeof(ptrId) {
    if (!ptrRegistry.has(ptrId)) return 0;
    const entry = ptrRegistry.get(ptrId);
    if (Buffer.isBuffer(entry.addr)) {
        return entry.addr.length;
    }
    return 0;
}

/**
 * pt_clear_all - Clear entire registry (for cleanup)
 */
function ptClearAll() {
    ptrRegistry.clear();
    nextPtrId = 1000;
}

// Export for Node.js
module.exports = {
    ptNew, ptAlloc,
    ptGet, ptSet,
    ptAdd, ptSub,
    ptEq, ptNull, ptIsNull,
    ptFree, ptTypeof, ptValid,
    ptSizeof, ptClearAll
};
